airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py

```diff
@@ -0,0 +1,394 @@
+
+        return response.json()
+
+    @retry_connection_handler(max_tries=5, factor=5)
+    @retry_after_handler(max_tries=3)
+    def get(self, url: str, params=None) -> Union[Mapping[str, Any], List[Mapping[str, Any]]]:
+        response = self._session.get(self.BASE_URL + url, params=self._add_auth(params))
+        return self._parse_and_handle_errors(response)
+
+    def post(self, url: str, data: Mapping[str, Any], params=None) -> Union[Mapping[str, Any], List[Mapping[str, Any]]]:
+        response = self._session.post(self.BASE_URL + url, params=self._add_auth(params), json=data)
+        return self._parse_and_handle_errors(response)
+
+
+class StreamAPI(ABC):
+    entity = None
+
+    more_key = None
+    data_path = "results"
+
+    page_filter = "offset"
+    page_field = "offset"
+
+    chunk_size = 1000 * 60 * 60 * 24  # TODO: use interval
+    limit = 100
+
+    def __init__(self, api: API, start_date: str = None, **kwargs):
+        self._api: API = api
+        self._start_date = pendulum.parse(start_date)
+
+    @abstractmethod
+    def list(self, fields) -> Iterable:
+        pass
+
+    def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
+        default_params = {"limit": self.limit, "properties": ",".join(self.properties.keys())}
+
+        params = {**default_params, **params} if params else {**default_params}
+
+        while True:
+            response = getter(params=params)
+            if response.get(self.data_path) is None:
+                raise RuntimeError("Unexpected API response: {} not in {}".format(self.data_path, response.keys()))
+
+            for row in response[self.data_path]:
+                yield row
+
+            # pagination
+            if "paging" in response:  # APIv3 pagination
+                if "next" in response["paging"]:
+                    params["after"] = response["paging"]["next"]["after"]
+                else:
+                    break
+            else:
+                if not response.get(self.more_key, False):
+                    break
+                if self.page_field in response:
+                    params[self.page_filter] = response[self.page_field]
+
+    def read_chunked(self, getter: Callable, params: Mapping[str, Any] = None):
+        params = {**params} if params else {}
+        now_ts = int(pendulum.now().timestamp() * 1000)
+        start_ts = int(self._start_date.timestamp() * 1000)
+
+        for ts in range(start_ts, now_ts, self.chunk_size):
+            end_ts = ts + self.chunk_size
+            params["startTimestamp"] = ts
+            params["endTimestamp"] = end_ts
+            logger.info(f"Reading chunk from {ts} to {end_ts}")
+            yield from self.read(getter, params)
+
+    @property
+    def properties(self) -> Mapping[str, Any]:
+        if not self.entity:
+            return {}
+
+        props = {}
+        data = self._api.get(f"/properties/v2/{self.entity}/properties")
```

Comment:
can you add context on what this is doing as a comment?

 I've looked at the API docs for a bit and it's not obvious to me what this is doing. it seems like it's adding the type of the property based on dynamically getting it from the Hubspot API? is this related to [this ticket](https://github.com/airbytehq/airbyte/issues/2219)? why wouldn't we know the type of the property ahead of time? 

This is bringing me to a different point: can you clarify as a docstring at the top of this class what the different APIs we are using are, and point to the API docs? This will enable me as the reader to know exactly which API docs to look at (Hubspot has multiple)

Reply:
most objects in Hubspot have attributes located in the properties field, there are two types:
- default (built-in, depends on the version of API)
- custom (user-defined)
because the exact list and types are dynamic we retrieve them at runtime. Technically we don't need to know type of attribute here, but for building a schema object we will need it (#2219 )